%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
\usepackage{listings}
\usepackage{color}
\definecolor{gray}{rgb}{0.5,0.5,0.5}
\lstset{
  language=Java,
  showstringspaces=false,
  columns=flexible,
  basicstyle={\scriptsize \ttfamily},
  numbers=none,
  stringstyle=\color{gray},
  breaklines=true,
  breakatwhitespace=true
}
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{jvenn: an interactive Venn diagram viewer.}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff2},                   % id's of addresses, e.g. {aff1,aff2}
   noteref={n1},                        % id's of article notes, if any
   email={Philippe.Bardou@toulouse.inra.fr}   % email address
]{\inits{PB}\fnm{Philippe} \snm{Bardou}}
\author[
   addressref={aff1},
   corref={aff1},                       % id of corresponding address, if any
   noteref={n1},                        % id's of article notes, if any
   email={Jerome.Mariette@toulouse.inra.fr}
]{\inits{JM}\fnm{J\'{e}r\^{o}me} \snm{Mariette}}
\author[
   addressref={aff1},
   email={Frederic.Escudie@toulouse.inra.fr}
]{\inits{FE}\fnm{Fr\'{e}d\'{e}ric} \snm{Escudi\'{e}}}
\author[
   addressref={aff1},
   email={Christophe.Djemiel@toulouse.inra.fr}
]{\inits{CD}\fnm{Christophe} \snm{Djemiel}}
\author[
   addressref={aff1,aff2},
   email={Christophe.Klopp@toulouse.inra.fr}
]{\inits{CK}\fnm{Christophe} \snm{Klopp}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Plate-forme bio-informatique Genotoul / MIA-T, INRA}, % university, etc
  \street{Borde Rouge},                     %
  \postcode{31326}                                % post or zip code
  \city{Castanet-Tolosan},                              % city
  \cny{France}                                    % country
}
\address[id=aff2]{%
  \orgname{Plate-forme SIGENAE / GenPhySE, INRA}, % university, etc
  \street{Borde Rouge},                     %
  \postcode{31326}                                % post or zip code
  \city{Castanet-Tolosan},                              % city
  \cny{France}                                    % country
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
\note[id=n1]{Equal contributor} % note, connected to author
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background} %if any
Venn diagrams are commonly used to present list comparison results. For example, 
biologists show differences between gene lists originating from different 
conditions this way. It is also a standard tool to display method comparison results 
in bioinformatics. When the number
of input lists exceeds four the diagram becomes difficult to read. Alternative layouts
and dynamic display features can improve its readability.

\parttitle{Results} %if any
jvenn is a new JavaScript library. It processes lists and produces Venn diagrams. 
It handles up to six input lists and presents results using classical or Edwards-Venn 
layouts. With it, developers can easily embed dynamic Venn diagrams in Web pages. 
jvenn allows to control and customize user interactions.

\parttitle{Conclusions} %if any
We introduce jvenn, an open source component for Web environments helping
scientists to analyze their data. The library package, which comes with full
documentation and an example, is freely available at
http://bioinfo.genotoul.fr/jvenn.


\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{Venn}
\kwd{Edward-Venn}
\kwd{vizualisation}
\kwd{jquery}
\kwd{JavaScript}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%

\section*{Background}

With the advent of high-throughput biology the number of compared samples,
within an experiment, is increasing. The analysis step often leads to the
production of a biological identifier list, such as gene names or operational
taxonomic units, for each sample. A common visualization chart is the Venn
diagrams \cite{Venn1880} which allows to spot shared and unshared identifiers
providing an insight on the similarities between the lists.

In a Venn diagram each list is figured by a transparent shape. Shape overlaps
contain the elements shared between lists or more often the corresponding counts.
In proportional Venn diagrams the size of a shape depends on the number of
elements of the corresponding list intersection. Venn diagram with up to four 
lists are easy to read and understand, but they become difficult to interpret
with more lists. To solve this problem, the Edwards-Venn \cite{Edwards2004}
representation introduces new shapes providing a clearer view (Fig. 2).

Many Venn diagram software packages are already available. The first six lines
of Table 1 present a subset of selected packages with their features including 
maximum number of input lists, input data formats, Venn diagram layouts,
application types and output formats. The table gives insight on several aspects
of Venn diagram production and highlights that no WEB applications handle up to
six lists. VENNTURE \cite{Bronwen2012} is the only application able to
produce such diagrams but this one only implements Edwards layout and only runs
under MS-Windows to produce static MS-PowerPoint files. Proportional Venn
diagrams are very limited in the number of lists they can display : maximum
three. Therefore programs designed to display high number of lists will
implement non proportional diagrams.

Hereafter we introduce jvenn, a JavaScript library, developed as a jQuery
plug-in \cite{jquery}, including many features easing diagram production and
enhancing their readability. The library has already been used and cited in two
scientific publications \cite{Bianchia2013, Aravindraja2013}. It is already
embedded in different Web applications such as nG6 \cite{Mariette2012},
RNAbrowse \cite{Mariette} and WallProtDB \cite{SanClemente}.


\section*{Implementation}

\subsection*{Inputs}

The library accepts three different input formats : lists, intersection counts
and count lists. All are JSON objects. Examples are presented in Table 1.
``Lists'' contain for each input a label and an identifier table. ``Intersection
counts'' contain a correspondence table between labels and letters [A..F] and a
table linking the intersection names formed by the successions of letters and
the counts. ``Count lists'' are organized as ``Lists'' in which identifiers are
replaced by their unique occurrences and their counts. With ``Count lists'' the 
figures presented in the diagram correspond to the sums of counts of all
identifiers shared between lists. This has been used in diversity studies to
present OTU (Operational Taxonomic Unit) read counts. For ``lists'' and ``count
lists'', jvenn computes the intersection counts and displays the chart. For 
``intersection counts'', it only displays the graphic. 

\subsection*{Display features}

Venn diagrams are commonly used to present up to six lists in its classical 
representation. But then, the intersection areas are often too small to display 
the figures.

To present, in a user-friendly manner, five or six list diagrams, jvenn implements
(Fig. 1) several functionalities. First, the display can be switched to Edwards-Venn
(Fig. 2) which gives a clearer graphical representation for six list diagrams. To
enhance the figure's readability on the classical six lists Venn graphic, it was
decided not to present all the values and to link some areas to their figures
using lines. This still did not permit to show all figures, therefore the switch
button panel (Fig. 1) was added. It enables to switch on and off the different lists
and display the corresponding intersection counts. For all diagrams, when the 
intersection count length exceeds the allowed space, the value is substituted by
a question mark. The value pops-up on mouse over. Last, to show the lists taking
part in an intersection, jvenn highlights the corresponding shapes on mouse
over, fading the others out.

The extra charts presented under the Venn diagram simplify verification and
comparison of multiple diagrams. The list size graph allows users to check the
homogeneity of the input list sizes. The intersection size graph can be used to
compare the compactness of multiple Venn diagrams.

For more than three lists diagrams, jvenn presents a switch button panel to 
highlight intersections (Fig. 1). It also provides two extra charts (Fig. 1) 
located bellow the Venn. The first one represents the input lists size
histogram. The second one displays the number of elements in intersections of a
certain size. It includes, as well, search and intersection identifiers export 
functions.

Scientists are usually interested in extracting identifier lists for some
intersections, therefore, jvenn implements an one-click function which retrieves
the names of the corresponding sets and the identifiers. To find an identifier
one can use the search box. The shapes containing the matching identifier are
then highlighted.

\subsection*{Outputs}

The display is based on a JavaScript canvas object allowing PNG export. The 
intersection table can also be downloaded as a CSV file. It contains a header 
line with the diagram area labels and, in column, the identifiers of the
elements contained in the area. This feature can be customized by the
developper.

\subsection*{Web application}

To simply produce a Venn diagram from identifier lists, jvenn is 
available as a Web application at http://bioinfo.genotoul.fr/jvenn/example.html
(Fig. 3).
The installation documentation is included in the software package which can be
downloaded from http://bioinfo.genotoul.fr/jvenn.

\subsection*{Performances}

jvenn's performance depends on the client browser. Using the running version
on a standard Linux computer (one cpu, 4GB of RAM), it displays a six lists
diagram of 10 000 identifiers in two seconds.


\section*{Results}

M.A. Dillies and colleagues \cite{Dillies2012} have compared seven RNA-Seq data
normalization methods and given a set of best practices to help biologists in their 
data processing. In table two, they have shown the differences between methods
pair-wise. The raw data table provided by the team contains 5,277 lines and
eight columns. The columns correspond to  the different methods presented in the
'Differential expression analysis' section of the article. The data in the table
was thresholded ($p < 0.05$) to produce the method specific gene name lists. Six
out of seven methods were selected for further processing ; Med was left out.
The lists were uploaded to the jvenn application and a Venn diagram was
produced. Using the layout selector the diagram was shown in Edwards Venn
format, in which all figures are accessible. This view presents all the lists
overlaps between methods. Considering Fig. 2, the higher values are located in
central areas of the graph showing that the methods share large portions of gene
lists. The list of 484 gene shared by DESeq, TMM, UQ and FQ has been extracted
by clicking on the corresponding figure. Gene G002562 was sought using the
search box. It was found to be part of the five genes shared by FQ and UQ. 

The jvenn statistics show that the different methods produce gene lists with
very different sizes (minimum 417 - maximum 1,249) and the most of the genes are
shared between methods : 1,069 genes out of 1,347 shared by at least four methods.

The same analysis was performed with VENNTURE, the only tool enabling to
generate a six list Edwards Venn diagram. First the software package was
installed on a computer running under MS-Windows. The six gene lists were loaded
in an MS-Excel spreadsheet. VENNTURE was run using the spreadsheet as input
generating a static MS-PowerPoint file containing the diagram and a MS-Excel
file with all the intersection contents. The names of 484 genes shared by DESeq,
TMM, UQ and FQ  were found in the intersection spreadsheet. The diagram did not
allow to search for gene G002562. Once more it was found in the intersection
spreadsheet.

\section*{Discussion}

jvenn enables to compare up to six lists and dynamically to update the diagram
by modifying the list content. Compared to VENNTURE it does not need any
installation and gives access to a dynamic diagram providing simple functions to
extract gene lists and perform searches. 

jvenns' statistical view gives a simple a quick overview of the sizes of the
different lists and of the overlaps. This enables to compare different Venn diagrams.

jvenn diagrams 

dynamic : list extraction, identifier search 

no statistical view 

Multiple layouts 



\section*{Conclusions}

jvenn allows whoever is familiar with the jquery library and has some JavaScript
programming skills to embed Venn diagrams in a Web page without dependency. It
has been designed following the examples of jbrowse \cite{Westesson01032013}, 
Cytoscape-Web \cite{Lopes2010}, and jHeatmap \cite{DeuPons2014}.

\section*{Availability and requirements}

jvenn is freely available under the GNU General Public License (GPL) and can be
downloaded with an example and the full documentation at
http://bioinfo.genotoul.fr/jvenn  website. A running version is accessible at
http://bioinfo.genotoul.fr/jvenn/example.html.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
The authors declare that they have no competing interests.

\section*{Author's contributions}
JM conceived and designed the project. JM, PB, FE and CD implemented the project.
CK evaluated software capabilities, and provided feedback on implementation. JM
and CK wrote the manuscript. All authors read and approved the final manuscript.

\section*{Acknowledgements}
We would like to acknowledge all our users for providing us useful feedback on
the system and for pointing out features worth developing. We thank the
reviewers for their insightful and constructive comments. 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file
\bibliography{bmc_jvenn}      % Bibliography file (usually '*.bib' )

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{A six lists classic Venn diagram.}
      This Venn diagram displays overlaps between six different biological
      samples. The icon, located on the top-right, allows users to download the
      diagram as a PNG file. The middle-right switch button
      panel allows to activate or dis-activate lists to access a specific
      intersection count. Charts showing the list size and intersection size 
      repartition located underneath the diagram.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{A six lists Edwards-Venn diagram.}
      On mouse over a figure, the shape corresponding to the lists involved in
      the intersection are highlighted and the other ones faded out. In
      this example, the user pointed the intersection between samples SRR068049,
      SRR068051 and SRR068052 which contains eight different items.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}

\begin{table}[h!]
\caption{Features of a subset of already available software packages, and jvenn.}
	\begin{tabular}{c|cccccc}
		Application & Maximum & Layouts & Application &
		Proportionality & Input data & Output\\ 
		& number of & & type & & formats & formats \\
		& input lists & & & & & \\ \hline
		
		VENNTURE \cite{Bronwen2012} &  6 & Edwards & Stand-alone & No & Lists &
		Powerpoint\\ 
		& & & & & & and Excel \\ \hline
		
		VennDiagram \cite{RVennDiagram} &  5 & Classical & R package & No & Lists
		& R object \\
		& & & & & & and TIFF \\ \hline
		
		BioVenn \cite{Hulsen2008} &  3 & Classical & Web application & Yes &
		Lists & SVG and PNG \\ \hline
		
		venny \cite{venny} &  4 & Classical & Web application & No &
		Lists & PNG \\ \hline
		 
		Canvasxpress \cite{canvasxpress} &  4 & Classical & JavaScript library &
		No & Intersection & JavaScript \\
		& & & & & counts & canvas \\ \hline
		
		Google & 3 & Classical & JavaScript library &
		Yes & Lists & PNG \\ 
		Chart API \cite{googleAPI} & & & & & & \\ \hline \hline
			
		jvenn & 6 & Classical & Web application & No & Lists,
		& Interactive 
		\\
		& & and Edwards & and JavaScript & & intersection & diagram, \\
		& & & library & & counts and & PNG and CSV \\
		& & & & & count lists & 
	\end{tabular}
\end{table}


\begin{table}[h!]
\caption{Available formats and example for the \textit{series} option.}
      \begin{tabular}{cccc}
        \hline
        format & example\\ \hline
        lists & 
\begin{lstlisting}
series: [{
	name: 'sample1',
	data: ["Otu1", "Otu2", "Otu3", "Otu4", "Otu5", "Otu6", "Otu7"]
}, {
	name: 'sample2',
	data: ["Otu1", "Otu2", "Otu5", "Otu7", "Otu8", "Otu9"]
}]
\end{lstlisting}\\ \hline
        intersection counts & 
\begin{lstlisting}
series: [{
	name: {A: 'sample 1', B: 'sample 2'},
	data: {A: ["Otu3", "Otu4", "Otu6"], B: ["Otu8", "Otu9"], AB: ["Otu1",
	"Otu2", "Otu5", "Otu7"]} }],
	values: {A: 3, B: 2, AB: 4}
\end{lstlisting}\\ \hline
        count lists  &
\begin{lstlisting}
series: [{
	name: 'sample1',
	data: ["Otu1", "Otu2", "Otu3", "Otu4", "Otu5", "Otu6", "Otu7"],
	values: [5, 15, 250, 20, 23, 58, 89]
}, {
	name: 'sample2',
	data: ["Otu1", "Otu2", "Otu5", "Otu7", "Otu8", "Otu9"],
	values: [90, 300, 10, 2, 45, 9]
}]
\end{lstlisting}\\ \hline
      \end{tabular}
\end{table}

\end{backmatter}
\end{document}